src/runtime-semantics/RegExp.mts

```diff
@@ -1,1187 +1,1307 @@
+    // const callResult = runMatcher(nextCall);
+    stack.push(matcher);
+    matcher = nextCall;
+    next = undefined;
   }
 }
+/** https://tc39.es/ecma262/#pattern-matcher */
+type Matcher = (x: MatchState, c: MatcherContinuation) => NonSpecFlattenedRegExpMatchingProcess;
+/** https://tc39.es/ecma262/#pattern-matchercontinuation */
+type MatcherContinuation = (y: MatchState) => NonSpecFlattenedRegExpMatchingProcess;
 
-type CharTester = (char: number) => boolean;
+type CharTester = (char: Character, canonicalize: RegExpRecord | undefined) => boolean;
+/** https://tc39.es/ecma262/#pattern-charset */
+abstract class CharSet {
+  abstract has(c: Character, rer: RegExpRecord | undefined): boolean;
 
-class UnionCharSet extends CharSet {
-  concrete;
+  abstract hasList(c: ListOfCharacter): boolean;
+
+  getStrings() {
+    return [...this.strings || []];
+  }
 
-  fns;
+  /**
+   * Return false if the Pattern is compiled in UnicodeSetMode and contains the empty sequence or sequences of more than one character.
+   */
+  abstract characterModeOnly: boolean;
 
-  constructor(concrete: Set<number>, fns: Set<CharTester>) {
-    super();
+  declare protected chars: Set<Character> | undefined;
+
+  declare protected strings: Set<ListOfCharacter> | undefined;
 
-    this.concrete = concrete;
-    this.fns = fns;
+  declare protected charTester: CharTester[] | undefined;
+
+  static union(...sets: CharSet[]) {
+    const unionChars = new Set<Character>();
+    const unionStrings = new Set<ListOfCharacter>();
+    let unionCharTesters: CharTester[] = [];
+    sets.forEach((set) => {
+      if (set.chars) {
+        set.chars.forEach((c) => unionChars.add(c));
+      }
+      if (set.strings) {
+        set.strings.forEach((s) => unionStrings.add(s));
+      }
+      if (set.charTester) {
+        unionCharTesters = unionCharTesters.concat(set.charTester);
+      }
+    });
+
+    if (!unionCharTesters.length) {
+      if (!unionStrings.size) {
+        return new ConcreteCharSet(unionChars);
+      }
+      if (!unionChars.size) {
+        return ConcreteStringSet.of(unionStrings);
+      }
+    }
+    if (!unionChars.size && !unionStrings.size && unionCharTesters.length === 1) {
+      return new VirtualCharSet(unionCharTesters[0]);
+    }
+    return new UnionCharSet(unionChars, unionStrings, unionCharTesters);
   }
 
-  has(c: number) {
-    if (this.concrete.has(c)) {
-      return true;
+  static intersection(...sets: CharSet[]): CharSet {
+    const intersectionChars = new Set<Character>();
+
+    const setChars = sets.filter((x) => x.chars);
+    setChars.sort((a, b) => a.chars!.size - b.chars!.size)[0]?.chars?.forEach((c) => {
+      if (setChars.every((s) => s.chars!.has(c))) {
+        intersectionChars.add(c);
+      }
+    });
+
+    const intersectionStrings = new Set<ListOfCharacter>();
+    const setStrings = sets.filter((x) => x.strings);
+    setStrings.sort((a, b) => a.strings!.size - b.strings!.size)[0]?.strings?.forEach((c) => {
```

Comment:
This is a little hard to read.  I think you're just looking for the smallest set size.  Using sort to do this is O(n log n), where a reduce might be O(n).

Of course, when n is small it doesn't really matter.  I'm more concerned about the readability of this code.